jvm.sh

```diff
@@ -1,66 +1,104 @@
-    new_java_home="$(grep "${version}"= ~/.jvmconfig | cut -f2 -d'=')"
-  elif [ -d "/usr/lib/jvm/java-${version}-oracle/" ]; then
-    new_java_home="/usr/lib/jvm/java-${version}-oracle/"
-  elif [ -e /usr/libexec/java_home ]; then
-    new_java_home="$(/usr/libexec/java_home -v 1."$version" || true)"
-  fi
-  if [ "$new_java_home" != "" ] && [ -d "$new_java_home" ]; then
-    if [ "$previous_java_home" != "" ] && [ "$previous_java_home" != "$new_java_home" ]; then
-      # shellcheck disable=SC2155
-      export PATH="$(echo "$PATH" | sed -e 's|'"$previous_java_home"'/bin:||g')"
-    fi
-    export JAVA_HOME="$new_java_home"
-    export PATH="${JAVA_HOME}/bin:$PATH"
-  fi
+  previous="$JAVA_HOME"
+
+  # custom jdk strategy
+  test -f ~/.jvmconfig && \
+    new="$(grep "$version"= ~/.jvmconfig || true | cut -f2 -d'=')"
+
+  # ubuntu/debian jdk strategy
+  test -z "$new" -a -d "/usr/lib/jvm/java-${version}-oracle/" && \
+    new="/usr/lib/jvm/java-${version}-oracle/"
+
+  # osx jdk strategy
+  test -z "$new" -a -e /usr/libexec/java_home && \
+    new="$(/usr/libexec/java_home -v 1."$version" || true)"
+
+  # sanity check: new must be a folder.
+  test ! -z "$new" -a -d "$new" || return 1
+
+  # PATH cleanup
+  # shellcheck disable=SC2155
+  test ! -z "$previous" -a "$previous" != "$new" && \
+    export PATH="$(echo "$PATH" | sed -e 's|'"$previous"'/bin:||g')"
+
+  # finally export new home and path
+  export JAVA_HOME="$new"
+  export PATH="${JAVA_HOME}/bin:$PATH"
 }
 
-# shellcheck disable=SC2039
-_jvm-discover-version() {
-  version=""
-  if [ -f pom.xml ]; then
-    version="$(\
-      grep '<java.version>' pom.xml | \
-      sed 's/.*<java.version>1.\(.*\)<\/java.version>.*/\1/' \
-    )"
-  fi
-  if [ -z "$version" ] && [ -f .java-version ]; then
-    version="$(cat .java-version)"
-  fi
-  if [ -z "$version" ] && [ -f ~/.java-version ]; then
-    version="$(cat ~/.java-version)"
-  fi
+# evaluates the 'maven.compiler.source' expression, returning the found java
+# version
+__jvm_pomversion_evaluate() {
+  MAVEN_OPTS="" mvn help:evaluate \
+    -Dexpression='maven.compiler.source' |
+    grep -e '^1\.[4-9]$' |
+    cut -f2 -d'.'
+}
+
+# tried to find the java version using regex.
+__jvm_pomversion_regex() {
+  grep -Eo '<(java.version|maven.compiler.source|source)>1\.[4-9]</.*>' pom.xml |
+    cut -f2 -d'>' |
+    cut -f2 -d'.' |
+    cut -f1 -d'<'
+}
+
+# tries multiple strategies to find the java version, and then sets it in a
+# .java-version
+__jvm_pomversion() {
+  test -f pom.xml || return 1
+  version="$(__jvm_pomversion_regex)"
+  test -z "$version" && version="$(__jvm_pomversion_evaluate)"
+  test ! -z "$version" && echo "$version" > .java-version
+}
+
+# finds out which java version should be used.
+__jvm_version() {
+  test ! -f .java-version && __jvm_pomversion
+  test -f .java-version && version="$(cat .java-version)"
+  test -z "$version" -a -f ~/.java-version && version="$(cat ~/.java-version)"
```

Comment:
Maybe isolate the last two lines in `__jvm_local_version()` and `__jvm_user_version()`? If those functions return 1 when not found, you could use

```bash
version=$(__jvm_local_version || __jvm_user_version)
```

Or even a future 

```bash
version=$(__jvm_local_version || __jvm_user_version || __jvm_system_version)
```

:)